### ISSUE TITLE
[synfusion_flutter_datagrid] DataGridController._dataGridPropertyChangeListeners is not removed when the widget is disposed

### ISSUE THREAD

**Issue by East333** [syncfusion/flutter-widgets#946]
there are two listeners added in SfDataGridState._setUp 
```dart
    _controller!._addDataGridPropertyChangeListener(
        _handleDataGridPropertyChangeListeners);

    _controller!
        ._addDataGridPropertyChangeListener(_handleSelectionPropertyChanged);
```
but they are not removed when the SfDataGridState disposed

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @East333,

We have checked your query at our end. We didn't remove the **_handleSelectionPropertyChanged** alone when the SfDataGridState is disposed of, the **_handleDataGridPropertyChangeListeners** was removed properly. We have considered this as a bug and logged a bug report in our feedback portal. We will fix the reported issue and include the changes in our upcoming 2022 Volume 3 SP release, which is expected to be rolled out on November 07, 2022 . We will update you once the release is rolled out, and we appreciate your patience until then. You can also track the status of the bug using the feedback link below.

Feedback Link: https://www.syncfusion.com/feedback/38710/the-datagridpropertychanged-listener-is-not-removed-properly-when-datagrid-is 

Regards,
Tamilarasan

**Comment by East333** [comment]
Unfortunately, you misunderstand my issue.Let me explain in detail.
The base Class of Class DataGridController is DataGridSourceChangeNotifier as follow:
```dart
class DataGridController extends DataGridSourceChangeNotifier {
  /// Creates the [DataGridController] with the [selectedIndex], [selectedRow]
  /// and [selectedRows].
  DataGridController(
      {int selectedIndex = -1,
      DataGridRow? selectedRow,
      List<DataGridRow> selectedRows = const <DataGridRow>[]})
      : _selectedRow = selectedRow,
        _selectedIndex = selectedIndex,
        _selectedRows = selectedRows.toList() {
    _currentCell = RowColumnIndex(-1, -1);
    _horizontalOffset = 0.0;
    _verticalOffset = 0.0;
  }
...
}
```
In the Class DataGridSourceChangeNotifier, there are some member variable and function such as follow:
```dart
  final ObserverList<_DataGridPropertyChangeListener>
      _dataGridPropertyChangeListeners =
      ObserverList<_DataGridPropertyChangeListener>();

  void _addDataGridPropertyChangeListener(
      _DataGridPropertyChangeListener listener) {
    _dataGridPropertyChangeListeners.add(listener);
  }

  void _removeDataGridPropertyChangeListener(
      _DataGridPropertyChangeListener listener) {
    _dataGridPropertyChangeListeners.remove(listener);
  }
```
when I use dataGridController.refreshRow, every listener in `dataGridController._dataGridPropertyChangeListeners` will be called.

In SfDataGridState.initState,` _handleDataGridPropertyChangeListeners` and `_handleSelectionPropertyChanged` are added in "dataGridController._dataGridPropertyChangeListeners", but after SfDataGridState.dispose , they are still in `dataGridController._dataGridPropertyChangeListeners` as follow:
```dart
  void _setUp() {
    _controller!._addDataGridPropertyChangeListener(
        _handleDataGridPropertyChangeListeners);


    _controller!
        ._addDataGridPropertyChangeListener(_handleSelectionPropertyChanged);
  }
```

Although, in SfDataGridState.dispose ` _controller?.removeListener(_handleDataGridPropertyChangeListeners);` will be called, but it can't remove `_handleDataGridPropertyChangeListeners` of `dataGridController._dataGridPropertyChangeListeners` and it's meaningless.

the situation is when The first SfDataGrid() is disposed,  I build a new second SfDataGrid() and use same dataGridController, when I call dataGridController.refreshRow, the listeners of first SfDataGrid() added in are not removed, but the state is dispost, so it will report an error such as "call setState after state disposed"


In summary, `dataGridController._dataGridPropertyChangeListeners` will not be removed in SfDataGridState.dispose, the code `_controller?.removeListener(_handleDataGridPropertyChangeListeners)` is useless, because the Listeners is not by  `_controller?.addListener` to add.


Looking forward to your reply,  thanks!

**Comment by East333** [comment]
Please review my code and issue.
Looking forward to your reply.

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @East333,

The fix for the reported issue has been included in the latest DataGrid version **20.3.57**. Update your DataGrid to get the issue resolved.

Pub Link: https://pub.dev/packages/syncfusion_flutter_datagrid/changelog
